src/main/asciidoc/java/index.adoc

```diff
@@ -0,0 +1,741 @@
+= Vert.x Codegen
+
+Vert.x Codegen is an annotation processing tool for processing Vert.x API and create API in different JVM lauguages.
+
+Vert.x polyglot langs use code generation for creating https://en.wikipedia.org/wiki/Shim_(computing)[shim] for APIs,
+that is thin wrappers in front of the API type.
+
+A shim is the adaptation of a Vert.x API in a JVM language. A shim uses code generation to generate
+the shim in its language from a Vert.x API, code generation is based on Java annotation processing.
+
+== Vert.x API
+
+A Vert.x API is a collection of Java types annotated with Codegen annotations that respect a set of constraints to
+ensure easy portability of the API to most languages and enable an asynchronous programming model. Such collection
+of API types are grouped in modules.
+
+=== Modules
+
+Modules are declared by annotating a Java package with a `link:../../apidocs/io/vertx/codegen/annotations/ModuleGen.html[@ModuleGen]`
+annotation.
+
+[source,java]
+----
+ModuleGen(name = "acme", groupPackage="com.acme")
+package com.acme.myservice;
+
+import io.vertx.codegen.annotations.ModuleGen;
+----
+
+The module _name_ is a namespace for shims that don't support package like naming, like JavaScript `acme-js`
+or Ruby `acme`, whereas the _group package_ is determines the created package, for instance
+for Groovy `com.acme.groovy.myservice`. The group package must be a prefix of the annotated module.
+
+Vert.x modules uses the reserved name _vertx-XYZ_ and the reserved group package _io.vertx_.
+
+An API module contains various Vert.x types annotated with `link:../../apidocs/io/vertx/codegen/annotations/VertxGen.html[@VertxGen]`
+or `link:../../apidocs/io/vertx/codegen/annotations/DataObject.html[@DataObject]`.
+
+NOTE: using Maven coordinates for name and group package is encouraged: the name corresponding to the
+Maven _artifactId_ and the group package corresponding to the `groupId`.
+
+=== Data objects
+
+_Data objects_ are Java class whose only purpose are to be data containers of various value types. A data object
+is mapped in some languages to Json.
+
+In its simplest form, a data object is a Java class following these rules:
+
+1. the class is annotated with `link:../../apidocs/io/vertx/codegen/annotations/DataObject.html[@DataObject]`
+2. provide a zero argument constructor
+3. provide a constructor with the `io.vertx.core.json.JsonObject` argument
+4. provide a copy constructor with the exact same type
+
+A data object can be also be an interface annotated with `@DataObject`, this is useful when multiple inheritance
+is needed. For instance Vert.x Core defines the `KeyCertOptions` and `TrustOptions` data object interfaces and the
+`JksOptions` is a data object class that implements both of them because a Java keystore provides support for both.
+
+Data object can also inherit other data objects, when it does so it inherits naturally its properties.
```

Comment:
inherit from other data objects. It inherits from the properties of the parent classes.